File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auto/Lift.java
package org.firstinspires.ftc.teamcode.auto;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

public class Lift {

    double ticksPerRotation = (((1+(46.0/17))) * (1+(46.0/11))) * 28;
    int mmPerRotation = 112;
    DcMotorEx motor;
    public Lift(HardwareMap hardwareMap) {
        motor = hardwareMap.get(DcMotorEx.class, "lift");
        motor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        motor.setDirection(DcMotorSimple.Direction.REVERSE);
    }

    public void setHeightMM(double heightMM) {
        int targetPosition = (int) (heightMM * ticksPerRotation / mmPerRotation);
        motor.setTargetPosition(targetPosition);
        motor.setMode(DcMotor.RunMode.RUN_TO_POSITION);

        if (motor.isBusy()) {
            motor.setPower(1);
        } else {
            motor.setPower(0);
            motor.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        }
    }

    public double getHeightMM() {
        return motor.getCurrentPosition() / ticksPerRotation * mmPerRotation;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/newteleop/systems/Lift.kt
package org.firstinspires.ftc.teamcode.newteleop.systems

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.robotcore.external.Telemetry
import kotlin.math.abs

class Lift(
    hardwareMap: HardwareMap,
    private val telemetry: Telemetry? = null
) {
    private val motor = hardwareMap.get(DcMotorEx::class.java, "lift")
    private val liftBottom = 50
    private val liftTop = 4080
    private val holdPower = 0.5
    private enum class LiftState {
        ADVANCING,
        BRAKING,
        HOLDING
    }
    private var state = LiftState.HOLDING
    private val minBrakeVelocity = 150
    var power: Number = 0.0
        set(value) {
            val clampedVal = value.toDouble().coerceIn(-1.0, 1.0)
            if (clampedVal != field) {
                if (field == 0.0) {
                    state = LiftState.BRAKING
                    motor.power = 0.0
                } else {
                    state = LiftState.ADVANCING
                    if (clampedVal > 0.0)
                        motor.targetPosition = liftTop
                    else
                        motor.targetPosition = liftBottom
                    motor.power = abs(clampedVal)
                }
                telemetry?.addData("Lift power", clampedVal)
                field = clampedVal
            }
        }

    init {
        motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        motor.direction = DcMotorSimple.Direction.REVERSE
        motor.targetPosition = 0
        motor.mode = DcMotor.RunMode.RUN_TO_POSITION
        motor.power = holdPower
    }

    fun update() {
        if (state == LiftState.BRAKING && motor.velocity <= minBrakeVelocity) {
            state = LiftState.HOLDING
            motor.targetPosition = motor.currentPosition
            motor.power = holdPower
        }
        telemetry?.addData("Lift position", motor.currentPosition)
        telemetry?.addData("Lift target position", motor.targetPosition)
        telemetry?.addData("Lift velocity", motor.velocity)
        telemetry?.addData("Lift state", state)
        telemetry?.addData("Lift mode", motor.mode)
        telemetry?.addData("Lift motor power", motor.power)
    }
}